### DISCUSSION TITLE
RiscV - SMP - LiteX / vexriscv_smp fails

### DISCUSSION THREAD

**Discussion by pottendo** [zephyrproject-rtos/zephyr#55049] [Q&A] [answered]
hi,
along my porting of Zephyr to a specific (not-yet-mainstream) FPGA board (see [here](https://github.com/pottendo/RVCop64-pottendo) and [here](https://github.com/pottendo/zephyr-pottendo)) I noticed that a vexriscv-smp softcore won't liftoff a Zephyr application if configured for SMP support (`CONFIG_SMP=y`). It crashes early in the boot process.
As the board is not widely available, I managed to get the issue reproduced using `litex_sim` for better analysis and fix hopefully.
The target hardware using a vexriscv_smp softcore in the FPGA runs linux/smp as expected; i.e. the OS detects 2 cores and multi-threaded applications speedup as expected.
However, due to memory limits of the board, Zephyr is by far better suited, so a working Zephyr SMP support would be a great feature.

Recently @npitre supported perfectly to make FPU working on this board, so I'm confident to fix SMP as well.

The problem has been analysed with one of the LiteX/VexRiscV experts (@Dolu1990) [here](https://github.com/enjoy-digital/litex/issues/1608).
RiscV / SMP Zephyr expert: Pls. have a look - it seems to be a an issue of the RiscV/SMP support within Zephyr in the area of spinlocks.

thanks, pottendo

PS: thanks for all the effort put into Zephyr - much appreciated!

**Comment by npitre** [comment, marked as answer]
You can't use CONFIG_ATOMIC_OPERATIONS_C with SMP.
You must use CONFIG_ATOMIC_OPERATIONSBUILTIN instead, which implies that
your CPU core must implement the A extension.

**Reply by Dolu1990** [reply]
Ahhh nice to know, thanks :)

**Reply by pottendo** [reply]
hi,
big thanx, SMP Zephyr apps lift-off now on litex_sim/vexriscv_smp CPUs!
`CONFIG_ATOMIC_OPERATIONS_C` was implicitly selected from the point where I started from: `SOC_RISCV32_LITEX_VEXRISCV`.
Cheers, pottendo

PS: removed time issue (was a stupid one).

**Comment by npitre** [comment]
`CONFIG_ATOMIC_OPERATIONS_C` is really a fallback if you can't do better even with SMP=n.

As for sleep not working, you should have a look in drivers/timer/riscv_machine_timer.c to be sure the base address and IRQ are the right ones.

**Reply by pottendo** [reply]
the timer is a no-issue - I forgot to adjust `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` properly for the 1MHz litex_sim.
